Adding resolveENS function and use on UserDetail Page

henrypalacios opened the pull request:
# Summary
Closes #869 

**Proposal:**
- Allow use `ENS` directly like URL as `/address/<ens_name>`

:warning: This solution prioritizes **Mainnet**, i.e. when searching `.../<network>/search/ensText.eth` it will search in **Mainnet**.

## To Test
- Visit `address` with valid :heavy_check_mark:  **ENS**
[/address/cosmoburn.eth](https://pr888--gpui.review.gnosisdev.com/address/cosmoburn.eth/)
- Visit `address` with Invalid :red_circle:  **ENS**
[/address/asereje.eth](https://pr888--gpui.review.gnosisdev.com/address/asereje.eth/)

alfetopito commented on src/apps/explorer/pages/UserDetails.tsx:
```diff
@@ -40,21 +44,48 @@ const TitleAddress = styled(RowWithCopyButton)`
+  const [resolvedAddress, setResolvedAddress] = useState<string | undefined | null>()
 
-  if (!isAddress(address)) {
+  useEffect(() => {
+    async function _resolveENS(): Promise<void> {
+      const _address = await resolveENS(address)
+      setResolvedAddress(_address)
+    }
+
+    if (isAddress(address)) {
+      setResolvedAddress(address)
+    } else if (isEns(address)) {
+      _resolveENS()
+    }
+  }, [address, networkId])
+
```
Not sure it would work, so feel free to ignore my suggestion.

What about creating a hook for this: `useResolveEns(address?:string) -> string | undefined` ?

henrypalacios replied:
I like it, I think it can work. Let me try...

alfetopito reviewed (approved):
Look great!

Just a minor suggestion

elena-zh commented:
is this OK that when I copy ENS name, I get an address instead?
https://watch.screencastify.com/v/fCf0QlwajPQq8cHHuFu0

henrypalacios commented:
@elena-zh I have added the possibility to copy ENS when it exists.

After discussing with @ramirotw  about the other networks I have updated the description of what I meant.

Although the **ENS** works for rinkeby, for example, [`rinkeby/address/agustin.eth`](https://pr888--gpui.review.gnosisdev.com/rinkeby/address/agustin.eth/) (our @alongoni) the resolved address will not be the same for [`mainnet/address/agustin.eth`](https://pr888--gpui.review.gnosisdev.com/address/agustin.eth/)

And in xDai, ENS does not seem to be [natively supported](https://www.xdaichain.com/for-developers/developer-resources/ens-compatibility).

elena-zh commented:
Hey @henrypalacios , @alfetopito , @ramirotw , as we are about to implement a network switcher functionality in the Explorer #160 , so it would be great to search for ENS mane with regard to selected network in the switcher. So I created a low prio enhancement #895  for this feature for the future.

elena-zh commented:
All the rest changes LGTM!

alfetopito commented:
> is this OK that when I copy ENS name, I get an address instead? https://watch.screencastify.com/v/fCf0QlwajPQq8cHHuFu0

Yeah, I am ok with that behaviour

alfetopito commented on src/hooks/useResolveEns.ts:
```diff
@@ -0,0 +1,31 @@
+    setResolvedEns(undefined)
+    if (isEns(address)) {
+      _resolveENS(address)
+    } else if (isAddress(address)) {
+      setResolvedEns({ address })
```
In this case this is not really an ENS address, is it?

I see what you are trying to do here but maybe the naming is misleading.

henrypalacios replied:
I have renamed the interface and state properties around [`AddressAccount`](https://github.com/gnosis/gp-ui/blob/869-address-page-by-ens/src/hooks/useResolveEns.ts#L6), what do you think?

alfetopito replied:
Looks better, thanks

henrypalacios commented on src/apps/explorer/pages/UserDetails.tsx:
```diff
@@ -40,21 +40,35 @@ const TitleAddress = styled(RowWithCopyButton)`
+          <h1>
+            User details
+            <TitleAddress
+              textToCopy={resolvedAddress.ens ? resolvedAddress.ens : resolvedAddress.address}
```
Here I have a doubt, I thought I understood that maybe you are OK with the first sharing of always copying the address, would it be OK to just leave [`addressAccount.address`](https://github.com/gnosis/gp-ui/blob/869-address-page-by-ens/src/apps/explorer/pages/UserDetails.tsx#L56)?

alfetopito replied:
Yep, that sounds good to me.

henrypalacios commented on src/hooks/useSearchSubmit.ts:
```diff
@@ -27,18 +39,7 @@ export function useSearchSubmit(): (query: string) => void {
-            })
-            .catch(() => history.push(`/${pathPrefix}/${query}`))
-        }
+        history.push(`/${path}/${query}`)
```
I have prioritized the search for **mainnet** through the search engine, when it is an ENS it will search in Mainnet, I would like to know your opinion about this @alfetopito.

Or would it be better to keep the network so that I can also comply with #895

alfetopito replied:
I think this is fine for now.

If this becomes a necessity we can do the query per network (if available) since it's not possible to guarantee a mainnet address exists on rinkeby/xdai networks.

elena-zh commented:
Just a tiny question: we say that orders can be found in another network when search by ENS name. 
But actually, we search by an address under this ENS name in the Mainnet. Taken into account that there is no registered `cosmoburn.eth` name in Rinkeby, it might seem that orders for this ENS name exist there. 

![image](https://user-images.githubusercontent.com/70885163/146174217-600f7c50-ca66-4c68-9349-36fd36ebe4a4.png)
Should we leave it as it is or we should not show that orders are found for this ENS name in another network? 

As another option, show a bit more explanative message `"No orders found on Mainnet for < address > ."` or so

![image](https://user-images.githubusercontent.com/70885163/146175298-61a833e4-cd45-4392-a5ba-cd533ed09956.png)

henrypalacios commented:
I like your suggestion Elena.

Seems to make sense to me, WDYT @alfetopito?

alfetopito commented:
> I like your suggestion Elena to add `for <address>`.
> 
> Seems to make sense to me, WDYT @alfetopito?

Agree!

henrypalacios commented:
I have added @elena-zh 's suggestion, I would like to know your opinions @alongoni, @alfetopito
![Selection_395](https://user-images.githubusercontent.com/4270166/146266995-61c6adc4-038a-4932-9c2a-90082f48bfa7.png)

alfetopito commented:
> I have added @elena-zh 's suggestion, I would like to know your opinions @alongoni, @alfetopito ![Selection_395](https://user-images.githubusercontent.com/4270166/146266995-61c6adc4-038a-4932-9c2a-90082f48bfa7.png)

I would display the address with the same component used in the title for consistency, with link to etherscan.
![Screenshot from 2021-12-16 08-03-01](https://user-images.githubusercontent.com/43217/146360369-de46df8c-df4a-4f2d-a3c5-3f932d4a2186.png)

elena-zh commented:
@henrypalacios , 
100% agree with @alfetopito 
Besides, it would be nice to make it look better in a small mobile screens
![image](https://user-images.githubusercontent.com/70885163/146361842-fd6962dd-befb-4b00-8a19-7f594ee2726c.png)

henrypalacios commented:
@alongoni Can you help me with this mismatch on small 
![Selection_399](https://user-images.githubusercontent.com/4270166/146390248-2349bfd3-6e43-445b-947c-e4270a8b6aae.png)
mobile devices?

henrypalacios commented:
Well done @alongoni, 

@ramirotw and @matextrem, please review this PR.

matextrem reviewed (commented):
LGTM 👍
